uqcsbot/advent.py

```diff
@@ -134,6 +138,111 @@
 }
 
 
+class LeaderboardView(discord.ui.View):
+    INITIAL_VISIBLE_ROWS = 20
+
+    def __init__(
+        self,
+        bot: UQCSBot,
+        code: int,
+        year: int,
+        day: Optional[Day],
+        members: list[Member],
+        leaderboard_style: str,
+        sortby: Optional[SortingMethod],
+    ):
+        super().__init__(timeout=300)
+        # constant within one embed
+        self.bot = bot
+        self.code = code
+        self.year = year
+        self.day = day
+        self.all_members = members
+        self.leaderboard_style = leaderboard_style
+        self.sortby = sortby
+        self.timestamp = datetime.now()
+
+        # can be changed by interaction
+        self.visible_members = members[: self.INITIAL_VISIBLE_ROWS]
+        self.show_text = False
+
+    @property
+    def is_truncated(self):
+        return len(self.visible_members) < len(self.all_members)
+
+    def make_message_arguments(self) -> Dict[str, Any]:
+        view_url = LEADERBOARD_VIEW_URL.format(year=self.year, code=self.code)
+
+        leaderboard = print_leaderboard(
+            parse_leaderboard_column_string(self.leaderboard_style, self.bot),
+            self.visible_members,
+            self.day,
+        )
+
+        title = (
+            "Advent of Code UQCS Leaderboard"
+            if self.code == UQCS_LEADERBOARD
+            else f"Advent of Code Leaderboard `{self.code}`"
+        )
+        title = f":star: {title} :trophy:"
+        if self.day:
+            title += f" \u2014 Day {self.day}"
+
+        notes: List[str] = []
+        if self.day:
+            notes.append(f"sorted by {self.sortby}")
+        if self.is_truncated:
+            notes.append(
+                f"top {len(self.visible_members)} shown out of {len(self.all_members)}"
+            )
+        body = f"({', '.join(notes)})" if notes else ""
+
+        basename = f"advent_{self.code}_{self.year}_{self.day}"
+
+        embed = discord.Embed(
+            title=title,
+            url=view_url,
+            description=body,
+            colour=discord.Colour.from_str(HL_COLOUR),
+            timestamp=self.timestamp,
+        )
+
+        if not self.show_text:
+            scoreboard_image = render_leaderboard_to_image(leaderboard)
+            file = discord.File(io.BytesIO(scoreboard_image), basename + ".png")
+            embed.set_image(url=f"attachment://{file.filename}")
+        else:
+            scoreboard_text = render_leaderboard_to_text(leaderboard)
+            file = discord.File(
+                io.BytesIO(scoreboard_text.encode("utf-8")), basename + ".txt"
+            )
+
+        self.show_all_interaction.disabled = not self.is_truncated
```

Comment:
It would be nice imo if this one could also be toggled back and forth, same as the text/image toggle